File: solutions/0/0/7/4/p0074_s0.c
/* Solution to Euler Problem 74: Digit Factorial Chains. */
#include "runner.h"

static const int digit_factorials[10] = {1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880};

/* Successor under the digit-factorial map: sum of factorials of n's decimal digits. */
static int sum_of_digit_factorials(int n) {
    int result = 0;
    while (n > 0) {
        result += digit_factorials[n % 10];
        n /= 10;
    }
    return result;
}

/* Walk-and-cache memoization over the digit-factorial functional graph: from each start, walk
 * forward until hitting a node with a known chain length or closing a cycle, then back-propagate
 * the length to every node on the walk (seen[i] gets total - i). A flat array of 3,000,000 slots
 * caches chain lengths and successors, since every reachable value stays below 7*9! = 2540160.
 * Amortised O(1) per start, so O(max_num) overall. */
const char *solve(int argc, char *argv[]) {
    static char _answer[32];
    int max_num = parse_int(argv[1]);

    /* We need a cache for chain lengths and graph (next number).
     * Numbers reachable from [1..max_num] are at most 6*9! = 2177280 for 7-digit numbers.
     * We'll use a hash map approach with open addressing. */

    /* Max reachable value: sum of digit factorials of any number up to 10M is at most 7*362880=2540160.
     * Use cache size of 3000000 to be safe. */
    int cache_size = 3000000;
    int *chain_length_cache = calloc((size_t)cache_size, sizeof(int));
    int *graph = calloc((size_t)cache_size, sizeof(int));
    if (!chain_length_cache || !graph) {
        fprintf(stderr, "Out of memory\n");
        { snprintf(_answer, sizeof _answer, "%lld", (long long)(-1)); return _answer; }
    }
    /* 0 means not cached */

    /* seen list for the walk */
    int seen_capacity = 200;
    int *seen = malloc((size_t)seen_capacity * sizeof(int));
    if (!seen) {
        fprintf(stderr, "Out of memory\n");
        { snprintf(_answer, sizeof _answer, "%lld", (long long)(-1)); return _answer; }
    }

    int max_chain_length = 0;
    int max_chain_length_count = 0;

    for (int start = 2; start <= max_num; start++) {
        int seen_len = 0;
        int current = start;

        /* Walk forward until we hit a cached node or revisit a node in this walk.
         * Cycle detection uses a linear scan of seen; chain lengths are bounded near 60,
         * so this constant-bounded scan beats the overhead of a hash set. */
        while (1) {
            /* Check if current is in seen (cycle detection) */
            int in_seen = 0;
            for (int i = 0; i < seen_len; i++) {
                if (seen[i] == current) {
                    in_seen = 1;
                    break;
                }
            }
            if (in_seen) break;

            /* Check if current is in cache */
            if (current < cache_size && chain_length_cache[current] != 0) break;

            /* Add to seen */
            if (seen_len >= seen_capacity) {
                seen_capacity *= 2;
                seen = realloc(seen, (size_t)seen_capacity * sizeof(int));
                if (!seen) {
                    fprintf(stderr, "Out of memory\n");
                    { snprintf(_answer, sizeof _answer, "%lld", (long long)(-1)); return _answer; }
                }
            }
            seen[seen_len++] = current;

            /* Compute next if not in graph */
            if (current < cache_size && graph[current] == 0) {
                graph[current] = sum_of_digit_factorials(current);
            }
            if (current < cache_size) {
                current = graph[current];
            } else {
                current = sum_of_digit_factorials(current);
            }
        }

        /* Chain length: walk length plus the cached tail, or just the walk if we closed a cycle. */
        int length;
        if (current < cache_size && chain_length_cache[current] != 0) {
            length = seen_len + chain_length_cache[current];
        } else {
            /* cycle: current is in seen */
            length = seen_len;
        }

        /* Back-propagate: seen[i] is i steps along the same chain, so its length is total - i. */
        for (int i = 0; i < seen_len; i++) {
            int num = seen[i];
            if (num < cache_size) {
                chain_length_cache[num] = length - i;
                /* graph[num] already set or set it now */
                if (graph[num] == 0) {
                    graph[num] = (i + 1 < seen_len) ? seen[i + 1] : current;
                }
            }
        }

        /* Check start's chain length */
        int cl = (start < cache_size) ? chain_length_cache[start] : length;
        if (cl > max_chain_length) {
            max_chain_length = cl;
            max_chain_length_count = 1;
        } else if (cl == max_chain_length) {
            max_chain_length_count++;
        }
    }

    free(chain_length_cache);
    free(graph);
    free(seen);

    { snprintf(_answer, sizeof _answer, "%lld", (long long)((long long)max_chain_length_count)); return _answer; }
}

File: solutions/0/0/7/4/p0074_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 74: Digit Factorial Chains [Level 2]. """
from __future__ import annotations

from solver.runners import runner

digit_factorials: tuple[int, ...] = (1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880)


def sum_of_digit_factorials(n: int) -> int:
    """Successor under the digit-factorial map: sum of factorials of n's decimal digits."""
    result = 0
    while n > 0:
        result += digit_factorials[n % 10]
        n //= 10
    return result


def find_max_length_chains_digit_factorial(max_num: int) -> tuple[int, int]:
    """Walk-and-cache memoization of chain lengths, then count those equal to the maximum."""
    chain_length_cache: dict[int, int] = {}
    graph: dict[int, int] = {}
    max_chain_length: int = 0
    max_chain_length_count: int = 0
    for start in range(2, max_num + 1):
        # Walk forward until we revisit a node in this walk (cycle) or hit a cached length.
        seen: list[int] = []
        current: int = start
        while current not in seen and current not in chain_length_cache:
            seen.append(current)
            if current not in graph:
                graph[current] = sum_of_digit_factorials(current)
            current = graph[current]
        if current in chain_length_cache:
            length = len(seen) + chain_length_cache[current]
        else:
            length = len(seen)
        # Back-propagate: seen[i] is i steps along the same chain, so its length is length - i.
        for i, num in enumerate(seen):
            chain_length_cache[num] = length - i
            graph[num] = seen[i + 1] if i + 1 < len(seen) else current
        if (chain_length := chain_length_cache[start]) > max_chain_length:
            max_chain_length = chain_length
            max_chain_length_count = 1
        elif chain_length == max_chain_length:
            max_chain_length_count += 1
    return (max_chain_length, max_chain_length_count)


@runner.main
def solve(*args: str) -> str:
    """Memoize chain lengths over the digit-factorial functional graph, back-propagating each
    walk's lengths into a cache so every start is amortised O(1); count starts hitting the
    maximum. Overall O(max_num)."""
    max_num = runner.parse_int(args[0])

    max_chain_length, max_chain_length_count = find_max_length_chains_digit_factorial(max_num)
    if runner.show:
        print(f"max_num={max_num!r} "
              f"max_chain_length={max_chain_length!r} "
              f"max_chain_length_count={max_chain_length_count!r}")
    return str(max_chain_length_count)


if __name__ == "__main__":
    raise SystemExit(solve())
